src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderRepository.cs

```diff
@@ -0,0 +1,378 @@
+                }
+
+                this.logger.LogTrace("(-)");
+            });
+
+            return task;
+        }
+
+        public Task<Dictionary<int, ProvenBlockHeader>> GetAsync(int fromBlockHeight, int toBlockHeight)
+        {
+            Task<Dictionary<int, ProvenBlockHeader>> task = Task.Run(() =>
+            {
+                this.logger.LogTrace("({0}:'{1}')", nameof(fromBlockHeight), fromBlockHeight);
+                this.logger.LogTrace("({0}:'{1}')", nameof(toBlockHeight), toBlockHeight);
+
+                using (DBreeze.Transactions.Transaction transaction = this.dbreeze.GetTransaction())
+                {
+                    var items = new Dictionary<int, ProvenBlockHeader>();
+
+                    transaction.SynchronizeTables(ProvenBlockHeaderTable);
+
+                    transaction.ValuesLazyLoadingIsOn = false;
+
+                    this.logger.LogTrace("Loading ProvenBlockHeaders from block height '{0}' to '{1}  from the database.",
+                        fromBlockHeight, toBlockHeight);
+
+                    for (int i = fromBlockHeight; i <= toBlockHeight; i++)
+                    {
+                        Row<byte[], ProvenBlockHeader> row =
+                            transaction.Select<byte[], ProvenBlockHeader>(ProvenBlockHeaderTable, i.ToBytes(false));
+
+                        if (row.Exists)
+                            items.Add(i, row.Value);
+                    }
+
+                    this.logger.LogTrace("(-)");
+
+                    return items;
+                }
+            });
+
+            return task;
+        }
+
+        /// <inheritdoc />
+        public Task<ProvenBlockHeader> GetAsync(int blockHeight)
+        {
+            Guard.NotNull(blockHeight, nameof(blockHeight));
+
+            Task<ProvenBlockHeader> task = Task.Run(() =>
+            {
+                this.logger.LogTrace("({0}:'{1}')", nameof(blockHeight), blockHeight);
+
+                using (DBreeze.Transactions.Transaction transaction = this.dbreeze.GetTransaction())
+                {
+                    ProvenBlockHeader header = null;
+
+                    transaction.SynchronizeTables(ProvenBlockHeaderTable);
+
+                    transaction.ValuesLazyLoadingIsOn = false;
+
+                    this.logger.LogTrace("Loading ProvenBlockHeader hash '{0}' from the database.", blockHeight);
+
+                    Row<byte[], ProvenBlockHeader> row =
+                        transaction.Select<byte[], ProvenBlockHeader>(ProvenBlockHeaderTable, blockHeight.ToBytes(false));
+
+                    if (row.Exists)
+                        header = row.Value;
+
+                    this.logger.LogTrace("(-)");
+
+                    return header;
+                }
+            });
+
+            return task;
+        }
+
+        /// <inheritdoc />
+        public Task PutAsync(List<ProvenBlockHeader> headers, HashHeightPair newTip)  
+        {
+            Guard.NotNull(headers, nameof(headers));
```

Comment:
I think it will be reasonable to assert that hash of the new tip is equal to the hash of the last provided header

Reply:
I thought the `HashHeightPair` hash is the block hash (blockId).  How would I derive this value within `ProvenBlockHeader` ?